Add matchingGesture and opposingGesture properties #trivial

SD10 opened the pull request:
Hey everyone @IBAnimatable 😃,
I'm pretty new to programming and open source but I'd like to start contributing to this project.
Let me know if there's anything specific I can do to help as I try to familiarize myself with the project.

### Summary of Pull Request:

adds `matchingGesture` and `opposingGesture` properties to `TransitionAnimationType.Direction`
adds `opposite` and `opposingGesture` properties to `TransitionAnimationType.HollowState`
adds `opposite` property to `TransitionAnimationType.PageType`

This helps condense the Animator initializers and prevent making a logic error using hard-coded values.

Removed unnecessary imports of `Foundation` where `UIKit` was already imported.

Fixed the name of `confingBarButtonItemImage()` -> `configBarButtonItemImage()`

SD10 committed:
Add matchingGesture and opposingGesture properties

tbaranes commented:
Thanks for doing this cleanup @SD10, welcome in the open source world 😬 
I really like the idea of `oposite`, seems to be a great move 👍 

I did a first quick look, it seems pretty cool but I will do a deeper look later.

As Danger reported, you have some [SwiftLint](https://github.com/realm/SwiftLint) warnings, if you want to have them directly in Xcode, you just have to install [SwiftLint](https://github.com/realm/SwiftLint).

Also, even if Danger said that you should add a CHANGELOG entry, I don't think that's necessary since it doesn't add any features, you can just add `#trivial` to your PR title in order to make Danger ignore that warning (even if it isn't trivial at all).

tbaranes commented on IBAnimatable/AnimatableBarButtonItem.swift:
```diff
@@ -45,6 +45,6 @@ open class AnimatableBarButtonItem: UIBarButtonItem, BarButtonItemDesignable, An
   // MARK: - Private
   fileprivate func configureInspectableProperties() {
 //    configureAnimatableProperties()
```
Can you remove that line? I know it's not related to your PR, but it still could be a good move!

SD10 replied:
Done

SD10 commented:
@tbaranes Thank you! 
I cleaned up the SwiftLint warnings and will have to install this to make it easier in the future.
I added the #trivial tag.

No worries if the whole PR is trivial. I love to refactor and condense code (while maintaining readability). The purpose was really to introduce myself and start to familiarize myself with the project.

tbaranes commented on IBAnimatable/TransitionAnimationType.swift:
```diff
@@ -190,6 +190,32 @@ extension TransitionAnimationType {
+      case .right: return .left
+      case .top: return .bottom
+      case .bottom: return .top
+      case .forward: return .close
```
`return .close` => `return .backward` 🙃

SD10 replied:
This property goes from 
`TransitionAnimationType.Direction` -> `InteractiveGestureType.GestureDirection`
`.forward` -> `.close`
`.backward` -> `.open`

To go the other way I would have to add a property to `InteractiveGestureType.GestureDirection` -> `TransitionAnimationType.Direction`

Which at the moment would be a superfluous property as we don't use this in the codebase

tbaranes replied:
Got it. Thanks for clarifying that point!

JakeLin replied:
Can you put your comment into the code as a `//blah blah` for future reference? That can help us remember what we did. It is very cool actually.

tbaranes commented on IBAnimatable/TransitionAnimationType.swift:
```diff
@@ -190,6 +190,32 @@ extension TransitionAnimationType {
+      case .top: return .bottom
+      case .bottom: return .top
+      case .forward: return .close
+      case .backward: return .open
```
`return .open` => `return .forward` 🙃

SD10 replied:
Same as above

tbaranes commented on IBAnimatable/TransitionAnimationType.swift:
```diff
@@ -202,6 +228,24 @@ extension TransitionAnimationType {
+      case .open: return .close
+      case .close: return .open
+      case .none:
+        fatalError("No opposing GestureDirection for HollowState.none")
```
What about just returning `.none` in this case? (for this one and all the others `opposite` `fatalError`)

SD10 replied:
I thought about this too but returning `.none` seemed counterintuitive as it's not an opposite.

I thought `fatalError` would help catch some programmer errors but I can change this to the requested functionality.

SD10 replied:
EDIT: 
The HollowState `opposite` property is the only one this would work for

There is no opposing gesture in `GestureDirection` for the `.none` case of `HollowState`

Also not all `TransitionAnimationType.Direction` cases have a matching or opposing InteractiveGestureType.GestureDirection

Basically, there's not a 1 : 1 mapping between these two types. 
So we use `fatalError` if someone is trying to use something that has no counterpart.

JakeLin replied:
I agree with @SD10 `fatalError` can help catch some programmer errors.

tbaranes reviewed (changes requested):
I let a few comments, pretty cool cleanup 👍

tbaranes reviewed (approved):
All good 👍 
Let's wait for @JakeLin review, then we will merge it.

JakeLin commented on IBAnimatable/AnimatableBarButtonItem.swift:
```diff
@@ -45,6 +45,6 @@ open class AnimatableBarButtonItem: UIBarButtonItem, BarButtonItemDesignable, An
   fileprivate func configureInspectableProperties() {
 //    configureAnimatableProperties()
-    confingBarButtonItemImage()
+    configBarButtonItemImage()
```
Can you change it to `configureBarButtonItemImage`, we changed all `config` to `configure` when we upgraded to Swift 3. this one was missed. ☹️

SD10 replied:
No problem, if I find any others I'll take care of them as well.

JakeLin commented on IBAnimatable/CardsAnimator.swift:
```diff
@@ -20,10 +20,10 @@ public class CardsAnimator: NSObject, AnimatedTransitioning {
+      self.reverseAnimationType = .cards(direction: direction.opposite)
     default:
       self.transitionAnimationType = .cards(direction: .forward)
       self.reverseAnimationType = .cards(direction: .backward)
```
Can we remove `switch` here?

SD10 replied:
See comment below

JakeLin commented on IBAnimatable/FadeAnimator.swift:
```diff
@@ -20,12 +20,9 @@ public class FadeAnimator: NSObject, AnimatedTransitioning {
-      self.reverseAnimationType = .fade(direction: .in)
+    case .in, .out:
+      self.transitionAnimationType = .fade(direction: direction)
+      self.reverseAnimationType = .fade(direction: direction.opposite)
```
Good cleanup

JakeLin commented on IBAnimatable/FlipAnimator.swift:
```diff
@@ -26,11 +26,11 @@ public class FlipAnimator: NSObject, AnimatedTransitioning {
+      self.interactiveGestureType = .pan(from: direction.opposingGesture)
       reverse = true
     default:
       self.transitionAnimationType = .flip(from: .left)
```
We can use `self.transitionAnimationType = .flip(from: direction)` here as well?

JakeLin commented on IBAnimatable/FoldAnimator.swift:
```diff
@@ -39,22 +39,12 @@ public class FoldAnimator: NSObject, AnimatedTransitioning {
-      self.reverseAnimationType = .fold(from: .top, folds: folds)
-      self.interactiveGestureType = .pan(from: .top)
-      reverse = true
+    switch direction {
```
I think we don't even need the `switch` here, the `default` case is pretty much `.left`.

JakeLin reviewed (approved):
👍 Very nice PR. Please have a look at my comments. 

@SD10 you have done a great job on cleaning up the direction. After you had cleaned them up, I felt like most of the `switch`s are redundant.  Can we remove them?

For some case, we have `case .right, top, .bottom` and `default`, because I put `case .left` within `default`. Is it possible to remove entire `switch`? If We can't, I think we can have something like
```
switch(direction) {
case .left, .right. .top, .bottom:
self.transitionAnimationType = .flip(from: direction)
self.reverseAnimationType = .flip(from: direction.opposite)
self.interactiveGestureType = .pan(from: direction.opposingGesture)
default: 
  fatalError("some error message")
}
```

How do you think?

JakeLin commented:
@SD10 Very nice cleanup, your cleanup enable us to remove or simplify the `switch`s. Can you have a look at my comments? And let me know how do you think about removing `switch`s.  Well done 👍

JakeLin commented:
@tbaranes thanks for reviewing the PR and onboarding @SD10 💪, do you think we can add some other tag like `#no-public-changes` since this PR is not `#trivial`. I don't know Danger well. I would like to follow what you are doing in that case.

SD10 commented:
@JakeLin Thanks for your feedback 😄

### About removing the default cases/switch statements:

I initially wanted to remove the switch statement completely but:

The `TransitionAnimationType.Direction` enum has 9 cases

I left the default case because I didn't know how you wanted to handle passing in a case that isn't supported such as `.cross`. I think your idea of throwing a `fatalError()` would work well.

Also part of what confused me:

I saw that on line `13` of `TurnAnimator` the `interactiveGestureType` property 
has a default value of `.pan(from: .horizontal)`
but in the `default` case of the initializer it is set to `.pan(from: .right)`

I didn't know if this was a typo or on purpose and did not want to make any breaking changes.

tbaranes commented:
Closing in favour of #427

JakeLin commented:
@SD10 thanks for following up all comments 👍. I can't remember why I put `.horizontal` in `public var interactiveGestureType: InteractiveGestureType? = .pan(from: .horizontal)` . Probably it was a mistake, should be `.pan(from: .right)` I recken.
